crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,254 @@
+
+    ```py
+    from enum import Enum, auto
+
+    class E(Enum):
+        A = auto()
+        B = auto()
+        C = auto()
+    ```
+
+- A "literal class type": a type representing a single known class (and none of its possible subclasses).
+
+- A "literal function type": a type that represents a single known function
+    (and excludes all other functions, even if they have the same signature).
+
+- A "literal module type": a type representing a single known module.
+
+Since it is known that all inhabitants of a given singleton type share the same memory address,
+singleton types in unions can be safely narrowed by identity,
+using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):
+
+```py
+def f(x: str | None):
+    if x is None:
+        ...  # x can be narrowed to `str`
+    else:
+        ...  # x can be narrowed to `None`
+```
+
+All singleton types are also sealed types, and closed types; nearly all are single-value types.
+(See below for definitions of these other concepts.)
+
+The number of singleton types in Python is theoretically infinite
+(for any given theoretical runtime object `x`, you could theorise a singleton type `Y` for which
+the sole inhabitant of `Y` would be the runtime object `x`). However, only a small fraction of
+the set of possible singleton types are recognised as being singleton types by red-knot's model.
+
+## Sealed types
+
+A sealed type is a type for which it is known
+that there is only a finite and pre-determined set of inhabitants at runtime.
+It follows from this that for every sealed type,
+there is also only a finite and pre-determined set of *subtypes* of that type.
+
+All singleton types are sealed types; however, not all sealed types are singleton types.
+
+Examples of sealed types (other than the singleton types listed above) are:
+
+- `bool`:
+
+    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
+    - The only proper subtypes[^2] of `bool` are `Literal[True]`, `Literal[False]`, and `Never`;
+        their union is exactly equal to `bool`.
+
+- Enums: consider the following enum class:
+
+    ```py
+    from enum import Enum, auto
+
+    class Foo(Enum):
+        X = auto()
+        Y = auto()
+        Z = auto()
+    ```
+
+    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
+        `Foo.X`, `Foo.Y` and `Foo.Z`.
+
+    - The only proper subtypes of `Foo` are:
+
+        - `Literal[Foo.X]`
+        - `Literal[Foo.Y]`
+        - `Literal[Foo.Z]`
+        - `Literal[Foo.X, Foo.Y]`
+        - `Literal[Foo.X, Foo.Z]`
+        - `Literal[Foo.Y, Foo.Z]`
+        - `Never`
+
+        The union of the proper subtypes of `Foo` (`Literal[Foo.X, Foo.Y, Foo.Z] | Never`)
+        is exactly equal to the type `Foo`.
+
+Because a sealed type is equivalent to the union of all of its proper subtypes,
+for any given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
```

Comment:
```suggestion
for any given sealed type `X` where the only inhabitants of `X` are `A`, `B` and `C`,
```

Reply:
No, I think that would make the sentence inaccurate. Your phrasing is true for the specific case of enum members, which I use as an example in the sentences following this, but is not true in the abstract, and this sentence describes an abstract relation. Here I theorise a hypothetical sealed type where there are exactly three proper subtypes and describe how such a sealed type might work.

I think it's possibly confusing that I'm using the same letters in my abstract description of the relation as I then use in my enum example immediately below, since the enum example does not in fact correspond exactly to the abstract relation I'm describing here